src/utils/csv.ts

```diff
@@ -0,0 +1,133 @@
+import { Aggregate, DatapointsGetAggregateDatapoint } from '@cognite/sdk';
+import moment from 'moment-timezone';
+import { getGranularityInSeconds } from './utils';
+
+export interface DatapointsToCSVProps {
+  data: DatapointsGetAggregateDatapoint[];
+  granularity: string;
+  aggregate: Aggregate;
+  delimiter: string;
+  format?: string;
+}
+
+export function datapointsToCSV({
+  data,
+  granularity,
+  aggregate = 'average',
+  delimiter = ',',
+  format,
+}: DatapointsToCSVProps): string {
+  const separator = delimiter === 'tab' ? '\t' : delimiter;
+  const ids = data.map(({ id, externalId }) => externalId || id.toString());
+  const rest = generateCSVStrings({
+    data,
+    granularity,
+    aggregate,
+    delimiter: separator,
+    format,
+  });
+
+  return [['timestamp', ...ids].join(separator), ...rest].join('\n');
+}
+
+export function downloadCSV(src: string, filename: string = 'noname.csv') {
+  const blob = new Blob([src], { type: 'text/csv;charset=utf-8;' });
+
+  if (navigator.msSaveBlob) {
+    navigator.msSaveBlob(blob, filename);
+  } else {
+    const link = document.createElement('a');
+
+    if (link.download !== undefined) {
+      const url = URL.createObjectURL(blob);
+
+      link.setAttribute('href', url);
+      link.setAttribute('download', filename);
+      link.style.visibility = 'hidden';
+
+      document.body.appendChild(link);
+      link.click();
+      document.body.removeChild(link);
+    }
+  }
+}
+
+function getStartTimestampAndIterationNumber(
+  data: DatapointsGetAggregateDatapoint[]
+): [number, number] {
+  let startTimestamp = Infinity;
+  const iterationNumber = data.reduce((acc, { datapoints }) => {
+    const timestamp = datapoints[0].timestamp.getTime();
+
+    startTimestamp = startTimestamp > timestamp ? timestamp : startTimestamp;
+
+    return datapoints.length + acc;
+  }, 0);
+
+  return [startTimestamp, iterationNumber];
+}
+
+function generateCSVStrings({
+  data,
+  aggregate,
+  delimiter,
+  format,
+  granularity: granularityString = '5s',
+}: DatapointsToCSVProps): string[] {
+  const strings = [];
+  const granularity = getGranularityInSeconds(granularityString);
+
+  if (!granularity) {
+    return [];
+  }
+
+  const pointers = Array(data.length).fill(0);
+  const [startTimestamp, iterationNumber] = getStartTimestampAndIterationNumber(
+    data
+  );
+
+  let iteration = 0;
+  let iterationLeft = iterationNumber;
+  const step = granularity * 1000;
+
+  while (iteration <= iterationNumber && iterationLeft) {
+    let valuesGet = 0;
```

Comment:
```suggestion
    let hasValues = false;
```
this doesn't need to be a number

Reply:
It has to be a number because we use it to decrease number of iteration arranging timestamps

Reply:
I guess it was your initial idea, but you don't actually use it anywhere except `valuesGet`